File: projects/fullstack-image-converter/apps/ios/ConvertMate/Engine/ConversionQueue.swift
// MARK: - ConversionQueue.swift
// Concurrency-controlled batch runner using Swift async/await + TaskGroup.
// Mirrors packages/core ConversionQueue.

import Foundation

actor ConversionQueue {
    private let imageEngine = ImageConversionEngine()
    private let videoEngine = VideoConversionEngine()

    func runAll(
        jobs: [ConversionJob],
        options: ConversionOptions,
        onUpdate: @escaping (ConversionJob) -> Void
    ) async -> [ConversionJob] {
        await withTaskGroup(of: ConversionJob.self) { group in
            // Use a semaphore-like pattern via task slots
            let semaphore = AsyncSemaphore(limit: options.concurrency)

            for job in jobs {
                group.addTask {
                    await semaphore.wait()
                    defer { Task { await semaphore.signal() } }

                    var updated = job
                    updated.status = .processing(progress: 0)
                    onUpdate(updated)

                    do {
                        let engine = self.pickEngine(for: job)
                        guard let engine else {
                            updated.status = .error("Unsupported: \(job.file.inputFormat) → \(job.outputFormat)")
                            onUpdate(updated)
                            return updated
                        }

                        let resultURL = try await engine.convert(job: job, options: options) { p in
                            var inProgress = job
                            inProgress.status = .processing(progress: p)
                            onUpdate(inProgress)
                        }
                        updated.status = .done
                        updated.resultURL = resultURL
                        onUpdate(updated)
                    } catch {
                        updated.status = .error(error.localizedDescription)
                        onUpdate(updated)
                    }
                    return updated
                }
            }

            var results: [ConversionJob] = []
            for await result in group { results.append(result) }
            return results
        }
    }

    private func pickEngine(for job: ConversionJob) -> ConversionEngine? {
        let input = job.file.inputFormat
        let output = job.outputFormat
        if imageEngine.canConvert(from: input, to: output) { return imageEngine }
        if videoEngine.canConvert(from: input, to: output) { return videoEngine }
        return nil
    }
}

// MARK: - AsyncSemaphore (Swift concurrency, no DispatchSemaphore)

actor AsyncSemaphore {
    private var count: Int
    private var waiters: [CheckedContinuation<Void, Never>] = []

    init(limit: Int) { self.count = limit }

    func wait() async {
        if count > 0 { count -= 1; return }
        await withCheckedContinuation { waiters.append($0) }
    }

    func signal() {
        if waiters.isEmpty { count += 1 }
        else { waiters.removeFirst().resume() }
    }
}

File: projects/fullstack-media-converter/apps/android/app/src/main/java/com/convertmate/conversion/ConversionQueue.kt
package com.convertmate.conversion

import android.content.Context
import com.convertmate.model.ConversionJob
import com.convertmate.model.ConversionOptions
import com.convertmate.model.JobStatus
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.sync.Semaphore
import kotlinx.coroutines.sync.withPermit

/**
 * Concurrency-controlled batch queue.
 * Mirrors packages/core ConversionQueue — same design, Kotlin coroutines instead of JS promises.
 */
class ConversionQueue(
    private val context: Context,
    private val imageEngine: ImageConversionEngine = ImageConversionEngine(),
    private val videoEngine: VideoConversionEngine = VideoConversionEngine(),
) {
    suspend fun runAll(
        jobs: List<ConversionJob>,
        options: ConversionOptions,
        onJobUpdate: (ConversionJob) -> Unit,
    ): List<ConversionJob> = coroutineScope {
        val semaphore = Semaphore(options.concurrency)
        jobs.map { job ->
            async {
                semaphore.withPermit {
                    onJobUpdate(job.copy(status = JobStatus.PROCESSING, progress = 0))
                    val engine = pickEngine(job.inputFormat, job.outputFormat)
                    if (engine == null) {
                        val err = job.copy(status = JobStatus.ERROR, error = "Unsupported: ${job.inputFormat} → ${job.outputFormat}")
                        onJobUpdate(err); err
                    } else {
                        val result = engine.convert(context, job, options) { p ->
                            onJobUpdate(job.copy(status = JobStatus.PROCESSING, progress = p))
                        }
                        onJobUpdate(result)
                        result
                    }
                }
            }
        }.awaitAll()
    }

    private fun pickEngine(input: String, output: String): ConversionEngine? = when {
        imageEngine.canConvert(input, output) -> imageEngine
        videoEngine.canConvert(input, output) -> videoEngine
        else -> null
    }
}
